pkg/ccl/crosscluster/logical/logical_replication_job.go

```diff
@@ -588,20 +591,56 @@ func loadOnlineReplicatedTime(
 }
 
 // OnFailOrCancel implements jobs.Resumer interface
-func (h *logicalReplicationResumer) OnFailOrCancel(
+func (r *logicalReplicationResumer) OnFailOrCancel(
 	ctx context.Context, execCtx interface{}, _ error,
 ) error {
 	execCfg := execCtx.(sql.JobExecContext).ExecCfg()
 	metrics := execCfg.JobRegistry.MetricsStruct().JobSpecificMetrics[jobspb.TypeLogicalReplication].(*Metrics)
 	metrics.ReplicatedTimeSeconds.Update(0)
+	r.completeProducerJob(ctx, execCfg.InternalDB)
 	return nil
 }
 
 // CollectProfile implements jobs.Resumer interface
-func (h *logicalReplicationResumer) CollectProfile(_ context.Context, _ interface{}) error {
+func (r *logicalReplicationResumer) CollectProfile(_ context.Context, _ interface{}) error {
 	return nil
 }
 
+func (r *logicalReplicationResumer) completeProducerJob(
+	ctx context.Context, internalDB *sql.InternalDB,
+) {
+	var (
+		progress = r.job.Progress().Details.(*jobspb.Progress_LogicalReplication).LogicalReplication
+		payload  = r.job.Details().(jobspb.LogicalReplicationDetails)
+	)
+
+	streamID := streampb.StreamID(payload.StreamID)
+	log.Infof(ctx, "attempting to update producer job %d", streamID)
+	if err := timeutil.RunWithTimeout(ctx, "complete producer job", 30*time.Second,
+		func(ctx context.Context) error {
+			client, err := streamclient.GetFirstActiveClient(ctx,
+				append([]string{payload.SourceClusterConnStr}, progress.StreamAddresses...),
+				internalDB,
+				streamclient.WithStreamID(streamID),
+				streamclient.WithLogical(),
+			)
+			if err != nil {
+				return err
+			}
+			defer closeAndLog(ctx, client)
+			return client.Complete(ctx, streamID, false /* successfulIngestion */)
+		},
+	); err != nil {
+		log.Warningf(ctx, "error completing the source cluster producer job %d: %s", streamID, err.Error())
+	}
+}
+
+func closeAndLog(ctx context.Context, d streamclient.Dialer) {
```

Comment:
In the ingestion job we do streamclient.Client.close(). Why use the type Dialer here?

Reply:
The most honest answer is this code was copy/pasta from the PCR job.

If I try to recall why we wrote that code in this way, I think it was mostly just that we want to log the error so we put it in the function and then we used streamclient.Dialer so we could use it in multiple places that have different kinds of clients.